include/solvers/navier_stokes_scratch_data.h

```diff
@@ -769,64 +783,132 @@ class NavierStokesScratchData : public PhysicsScratchDataBase
+   *
+   * @param void_fraction_solution The void fraction calculated with one of the methods of
+   * the VoidFractionBase class.
+   */
+
+  template <typename VectorType>
+  void
+  calculate_void_fraction_at_particle_location(
+    const Quadrature<dim>                                &q_particles,
+    const typename DoFHandler<dim>::active_cell_iterator &void_fraction_cell,
+    const VectorType &void_fraction_solution)
+  {
+    FEValues<dim> fe_values_particles_void_fraction(
+      this->fe_values_void_fraction->get_fe(), q_particles, update_values);
+
+    fe_values_particles_void_fraction.reinit(void_fraction_cell);
 
-        const auto &void_fraction_dh_cell =
-          typename DoFHandler<dim>::cell_iterator(
-            *this->fe_values_void_fraction->get_cell(),
-            &void_fraction_dof_handler);
+    fe_values_particles_void_fraction.get_function_values(
+      void_fraction_solution, cell_void_fraction);
+  }
 
-        fe_values_particles_void_fraction.reinit(void_fraction_dh_cell);
+  /** @brief Calculates the properties of the fluid at the locations of the particles.
+   * At the moment, only constant properties within the same fluid are
+   * supported. When two fluids are present and VOF is used, the properties are
+   * calculated based on the filtered phase fraction interpolated at the
+   * location of the particles. These properties are used in the forces
+   * calculations in the VANS equations.
+   */
 
-        fe_values_particles_void_fraction.get_function_values(
-          void_fraction_solution, cell_void_fraction);
+  void
+  calculate_fluid_properties_at_particle_location()
+  {
+    density_at_particle_location.resize(number_of_particles);
+    kinematic_viscosity_at_particle_location.resize(number_of_particles);
+
+    if (gather_vof)
+      {
+        for (unsigned int particle_i = 0; particle_i < number_of_particles;
+             ++particle_i)
+          {
+            density_at_particle_location[particle_i] = calculate_point_property(
+              filtered_phase_values_at_particle_location[particle_i],
+              this->density_ref_0,
+              this->density_ref_1);
+            double dynamic_viscosity_at_particle_location =
+              calculate_point_property(
+                filtered_phase_values_at_particle_location[particle_i],
+                this->kinematic_viscosity_scale_0 * this->density_ref_0,
+                this->kinematic_viscosity_scale_1 * this->density_ref_1);
+
+            kinematic_viscosity_at_particle_location[particle_i] =
+              dynamic_viscosity_at_particle_location /
+              density_at_particle_location[particle_i];
+          }
       }
+    else
+      {
+        for (unsigned int particle_i = 0; particle_i < number_of_particles;
+             ++particle_i)
+          {
+            // Calculate the kinematic viscosity and density at the particle
```

Comment:
"Get" or "calculate"?

Reply:
Gather would be the adequate here I think